File: core/parsing.go
package core

import (
	"encoding/hex"
	"errors"
	"fmt"
	"math/big"
	"reflect"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
)

// ParseTransactionData parses the transaction data and identifies the function call
func ParseTransactionData(to string, data string, chainID uint64, options VerifyOptions) (*CallData, error) {
	// Remove 0x prefix if present
	cleanData := strings.TrimPrefix(data, "0x")

	// Normalize the target address
	normalizedTo := strings.ToLower(to)

	// Check if this is a known contract
	contractInfo, isKnownContract := GetKnownContract(normalizedTo, chainID)
	targetName := ""
	if isKnownContract {
		targetName = contractInfo.Name
	}

	// If data is empty, return a simple transfer call
	if len(cleanData) == 0 {
		return &CallData{
			Target:       to,
			TargetName:   targetName,
			FunctionName: "unknown",
			RawData:      "0x" + data,
		}, nil
	}

	// Extract function selector (first 4 bytes)
	if len(cleanData) < 8 {
		return &CallData{
			Target:       to,
			TargetName:   targetName,
			FunctionName: "unknown",
			RawData:      data,
		}, nil
	}

	functionSelector := cleanData[:8]

	// Try to identify the function from known selectors
	functionInfo, isKnownFunction := KnownFunctions[functionSelector]

	if !isKnownFunction {
		// If we can't identify the function, return the raw data
		return &CallData{
			Target:       to,
			TargetName:   targetName,
			FunctionName: "unknown",
			RawData:      data,
		}, nil
	}

	// Parse the function arguments
	parsedArgs, err := parseArguments(functionInfo.ABI, "0x"+cleanData)
	if err != nil {
		// If we can't parse the arguments, return the raw data
		return &CallData{
			Target:       to,
			TargetName:   targetName,
			FunctionName: functionInfo.Name,
			RawData:      data,
		}, nil
	}

	// If this is a token function on a known contract where we have decimals, adjust the amount
	if isKnownContract && TokenFunctions[functionInfo.Name] && contractInfo.Decimals > 0 && parsedArgs["amount"] != nil {
		parsedArgs["amount"] = ParseDecimals(parsedArgs["amount"].(*big.Int), contractInfo.Decimals)
	}

	// Check if any of the arguments are known contracts
	for key, value := range parsedArgs {
		if value, ok := value.(common.Address); ok {
			contractInfo, isKnownContract := GetKnownContract(value.Hex(), chainID)
			if isKnownContract {
				parsedArgs[key] = fmt.Sprintf("%s (%s 🔍)", value, contractInfo.Name)
			}
		}
	}

	// Check if this is a multicall contract and the function is a multicall function
	isMulticallContract := false
	if chainMulticalls, exists := MulticallAddresses[chainID]; exists {
		isMulticallContract = chainMulticalls[normalizedTo]
	}

	isMulticallFunction := functionInfo.Name == "multiSend" || functionInfo.Name == "aggregate3" || functionInfo.Name == "aggregate3Value"

	if isMulticallContract && isMulticallFunction {
		// Parse subcalls - passing contract address, chain ID, and full function info
		subcalls, err := parseMulticall(normalizedTo, chainID, functionInfo, parsedArgs, options)
		if err != nil {
			return nil, err
		}

		// Return with subcalls
		return &CallData{
			Target:       to,
			TargetName:   targetName,
			FunctionName: functionInfo.Name,
			SubCalls:     subcalls,
		}, nil
	}

	// Regular function call
	return &CallData{
		Target:       to,
		TargetName:   targetName,
		FunctionName: functionInfo.Name,
		ParsedData:   parsedArgs,
	}, nil
}

// ParseDecimals parses the amount and returns it as a human-readable string
// with the correct number of decimals and comma grouping in the integer portion.
// If the decimal portion is all zeros, it will show only 2 decimal places.
func ParseDecimals(amount *big.Int, decimals int) string {
	amountStr := amount.String()

	// Pad with leading zeros for decimal portion if needed
	if len(amountStr) <= decimals {
		amountStr = strings.Repeat("0", decimals-len(amountStr)+1) + amountStr
	}
	decimalPos := len(amountStr) - decimals

	// Split integer and fractional parts
	intPart := amountStr[:decimalPos]
	fracPart := ""
	if decimals > 0 {
		fracPart = amountStr[decimalPos:]
	}

	// Insert commas into integer part
	intPartWithCommas := addCommas(intPart)

	// If we have a fractional part
	if decimals > 0 {
		// Check if all digits in fracPart are zeros
		allZeros := true
		for _, c := range fracPart {
			if c != '0' {
				allZeros = false
				break
			}
		}

		if allZeros {
			// If all zeros, just show 2 decimal places
			if len(fracPart) >= 2 {
				return intPartWithCommas + "." + fracPart[:2]
			}
			return intPartWithCommas + "." + fracPart + strings.Repeat("0", 2-len(fracPart))
		} else {
			// Trim trailing zeros
			for len(fracPart) > 0 && fracPart[len(fracPart)-1] == '0' {
				fracPart = fracPart[:len(fracPart)-1]
			}
			if len(fracPart) > 0 {
				return intPartWithCommas + "." + fracPart
			}
		}
	}

	return intPartWithCommas
}

// addCommas inserts commas every three digits starting from the right.
func addCommas(s string) string {
	// Special case for empty or very short string
	if len(s) <= 3 {
		return s
	}

	var result []byte
	count := 0

	// Append digits in reverse order, inserting commas every 3 digits
	for i := len(s) - 1; i >= 0; i-- {
		result = append(result, s[i])
		count++
		if count%3 == 0 && i != 0 {
			result = append(result, ',')
		}
	}

	// Reverse the result back to normal order
	for i := 0; i < len(result)/2; i++ {
		result[i], result[len(result)-1-i] = result[len(result)-1-i], result[i]
	}

	return string(result)
}

// parseArguments decodes the function arguments from calldata
func parseArguments(method abi.Method, calldata string) (map[string]interface{}, error) {
	// Remove 0x prefix if present
	calldata = strings.TrimPrefix(calldata, "0x")

	// Convert hex string to bytes
	data, err := hex.DecodeString(calldata)
	if err != nil {
		return nil, err
	}

	// Ensure data is long enough to contain the method ID
	if len(data) < 4 {
		return nil, errors.New("calldata too short")
	}

	// Skip the method ID (first 4 bytes)
	inputData := data[4:]

	// Unpack the arguments
	args, err := method.Inputs.Unpack(inputData)
	if err != nil {
		return nil, err
	}

	// Convert to a map
	result := make(map[string]interface{})
	for i, arg := range args {
		name := method.Inputs[i].Name
		if name == "" {
			name = fmt.Sprintf("arg%d", i)
		}

		// Handle arrays by creating indexed entries in the map
		if reflect.TypeOf(arg).Kind() == reflect.Slice &&
			reflect.TypeOf(arg).Elem().Kind() != reflect.Uint8 {
			// It's an array but not a byte array
			arrayValue := reflect.ValueOf(arg)

			// Add the full array under its original name
			result[name] = arg

			// Also add individual elements with indexed names for easier access
			for j := 0; j < arrayValue.Len(); j++ {
				indexedName := fmt.Sprintf("%s[%d]", name, j)
				element := arrayValue.Index(j).Interface()

				// Apply the same formatting we do for regular elements
				if byteArray, ok := element.([]byte); ok {
					result[indexedName] = "0x" + hex.EncodeToString(byteArray)
				} else {
					result[indexedName] = element
				}
			}
		} else {
			// Convert byte arrays and fixed-size uint8 arrays to hex strings for better readability
			if byteArray, ok := arg.([]byte); ok {
				result[name] = "0x" + hex.EncodeToString(byteArray)
			} else {
				// Check for fixed-size uint8 arrays using type string
				argType := fmt.Sprintf("%T", arg)
				if strings.HasPrefix(argType, "[") && strings.HasSuffix(argType, "]uint8") {
					// Convert the fixed-size array to a byte slice
					bytes := make([]byte, 0)

					// Get string representation without brackets
					str := fmt.Sprintf("%v", arg)
					str = strings.Trim(str, "[]")

					// Split by space (Go's array string representation uses spaces)
					for _, numStr := range strings.Fields(str) {
						var val uint8
						fmt.Sscanf(numStr, "%d", &val)
						bytes = append(bytes, val)
					}

					result[name] = "0x" + hex.EncodeToString(bytes)
				} else {
					result[name] = arg
				}
			}
		}
	}

	return result, nil
}

// parseMulticall parses subcalls from a multicall function
func parseMulticall(contractAddress string, chainID uint64, functionInfo FunctionInfo, args map[string]interface{}, options VerifyOptions) ([]CallData, error) {
	var subcalls []CallData

	// First determine which contract we're dealing with
	normalizedAddress := strings.ToLower(contractAddress)

	// Handle Safe Multisend contracts
	if normalizedAddress == strings.ToLower(SafeMultisendAddress) || normalizedAddress == strings.ToLower(SafeMultisendCallOnly141) {
		// For Safe Multisend contract, check the function signature
		if functionInfo.Signature == SafeMultisendSig {
			// Parse multiSend calldata
			hexData, ok := args["transactions"].(string)
			if !ok || !strings.HasPrefix(hexData, "0x") {
				return nil, errors.New("invalid multiSend data format")
			}

			// Convert hex string to bytes
			data, err := hex.DecodeString(strings.TrimPrefix(hexData, "0x"))
			if err != nil {
				return nil, fmt.Errorf("invalid multiSend data: %v", err)
			}

			// multiSend data format: operation (1 byte) + to (20 bytes) + value (32 bytes) + dataLength (32 bytes) + data (variable)
			pos := 0
			for pos < len(data) {
				// Ensure we have enough data for the fixed-size fields
				if pos+85 > len(data) {
					break
				}

				// Extract operation
				// Can skip extracting the operation as it's always 0
				pos++

				// Extract to address
				to := common.BytesToAddress(data[pos : pos+20])
				pos += 20

				// Skip extracting value for now, can implement later if needed
				pos += 32

				// Extract data length
				dataLength := common.BytesToHash(data[pos : pos+32])
				pos += 32

				// Convert data length to int
				length := new(big.Int).SetBytes(dataLength[:]).Uint64()

				// Ensure we have enough data for the variable-size field
				if pos+int(length) > len(data) {
					break
				}

				// Extract call data
				callData := data[pos : pos+int(length)]
				pos += int(length)

				// Parse the subcall
				subcall, err := ParseTransactionData(to.Hex(), "0x"+hex.EncodeToString(callData), chainID, options)
				if err != nil {
					return nil, err
				}

				subcalls = append(subcalls, *subcall)
			}
		} else {
			return nil, fmt.Errorf("unsupported function %s for Safe Multisend contract", functionInfo.Signature)
		}
	} else if normalizedAddress == strings.ToLower(Multicall3Address) || normalizedAddress == strings.ToLower(Multicall3Delegatecall) {
		// For Multicall3 contract, check the function signature
		if functionInfo.Signature == Aggregate3Sig {
			// Define the struct type for the calls
			type Call3 struct {
				Target       common.Address
				AllowFailure bool
				CallData     []byte
			}

			// Get and validate the raw calls data
			rawCalls, ok := args["calls"]
			if !ok {
				return nil, errors.New("missing calls in aggregate3 data")
			}

			// Use reflection to convert the data to our expected type
			rawCallsValue := reflect.ValueOf(rawCalls)
			if rawCallsValue.Kind() != reflect.Slice {
				return nil, errors.New("aggregate3 calls must be a slice")
			}

			// Create the result slice with the correct capacity
			calls := make([]Call3, rawCallsValue.Len())

			// Process each call in the slice
			for i := 0; i < rawCallsValue.Len(); i++ {
				callValue := rawCallsValue.Index(i)

				// Extract the required fields using case-insensitive matching
				targetField := callValue.FieldByNameFunc(func(name string) bool {
					return strings.EqualFold(name, "target")
				})
				if !targetField.IsValid() {
					return nil, errors.New("missing target field in call")
				}

				allowFailureField := callValue.FieldByNameFunc(func(name string) bool {
					return strings.EqualFold(name, "allowfailure")
				})
				if !allowFailureField.IsValid() {
					return nil, errors.New("missing allowFailure field in call")
				}

				callDataField := callValue.FieldByNameFunc(func(name string) bool {
					return strings.EqualFold(name, "calldata")
				})
				if !callDataField.IsValid() {
					return nil, errors.New("missing callData field in call")
				}

				// Populate the result struct
				calls[i] = Call3{
					Target:       targetField.Interface().(common.Address),
					AllowFailure: allowFailureField.Interface().(bool),
					CallData:     callDataField.Interface().([]byte),
				}
			}

			// Parse aggregate3 calldata
			for _, call := range calls {
				subcall, err := ParseTransactionData(call.Target.Hex(), "0x"+hex.EncodeToString(call.CallData), chainID, options)
				if err != nil {
					continue
				}

				subcalls = append(subcalls, *subcall)
			}
		} else if functionInfo.Signature == Aggregate3ValueSig {
			// Define the struct type for the calls
			type Call3 struct {
				Target       common.Address
				AllowFailure bool
				Value        *big.Int
				CallData     []byte
			}

			// Get and validate the raw calls data
			rawCalls, ok := args["calls"]
			if !ok {
				return nil, errors.New("missing calls in aggregate3 data")
			}

			// Use reflection to convert the data to our expected type
			rawCallsValue := reflect.ValueOf(rawCalls)
			if rawCallsValue.Kind() != reflect.Slice {
				return nil, errors.New("aggregate3 calls must be a slice")
			}

			// Create the result slice with the correct capacity
			calls := make([]Call3, rawCallsValue.Len())

			// Process each call in the slice
			for i := 0; i < rawCallsValue.Len(); i++ {
				callValue := rawCallsValue.Index(i)

				// Extract the required fields using case-insensitive matching
				targetField := callValue.FieldByNameFunc(func(name string) bool {
					return strings.EqualFold(name, "target")
				})
				if !targetField.IsValid() {
					return nil, errors.New("missing target field in call")
				}

				allowFailureField := callValue.FieldByNameFunc(func(name string) bool {
					return strings.EqualFold(name, "allowfailure")
				})
				if !allowFailureField.IsValid() {
					return nil, errors.New("missing allowFailure field in call")
				}

				valueField := callValue.FieldByNameFunc(func(name string) bool {
					return strings.EqualFold(name, "value")
				})
				if !valueField.IsValid() {
					return nil, errors.New("missing value field in call")
				}

				callDataField := callValue.FieldByNameFunc(func(name string) bool {
					return strings.EqualFold(name, "calldata")
				})
				if !callDataField.IsValid() {
					return nil, errors.New("missing callData field in call")
				}

				// Populate the result struct
				calls[i] = Call3{
					Target:       targetField.Interface().(common.Address),
					AllowFailure: allowFailureField.Interface().(bool),
					Value:        valueField.Interface().(*big.Int),
					CallData:     callDataField.Interface().([]byte),
				}
			}

			// Parse aggregate3 calldata
			for _, call := range calls {
				subcall, err := ParseTransactionData(call.Target.Hex(), "0x"+hex.EncodeToString(call.CallData), chainID, options)
				if err != nil {
					continue
				}

				subcalls = append(subcalls, *subcall)
			}
		} else {
			return nil, fmt.Errorf("unsupported function %s for Multicall3 contract", functionInfo.Signature)
		}
	} else {
		// No generic parsing - return an error for unknown contracts
		return nil, fmt.Errorf("unsupported multicall contract: %s", contractAddress)
	}

	return subcalls, nil
}

File: core/web/core/parsing.js
import { ethers } from 'ethers';
import { 
  SAFE_MULTISEND_ADDRESS,
  SAFE_MULTISEND_CALL_ONLY_141,
  SAFE_MULTISEND_SIG,
  MULTICALL3_ADDRESS,
  AGGREGATE3_SIG,
  KNOWN_FUNCTIONS as knownFunctions,
  TOKEN_FUNCTIONS as tokenFunctions,
  MULTICALL_ADDRESSES as multicallAddresses,
  getKnownContract
} from './constants.js';

/**
 * Parse the transaction data and identify the function call
 * @param {string} to - Target address
 * @param {string} data - Transaction data
 * @param {number} chainID - Chain ID
 * @param {Object} options - Verification options
 * @returns {Object} Parsed call data
 */
export function parseTransactionData(to, data, chainID, options = {}) {
  // Remove 0x prefix if present
  const cleanData = data.startsWith('0x') ? data.slice(2) : data;

  // Normalize the target address
  const normalizedTo = to.toLowerCase();

  // Check if this is a known contract
  const contractInfo = getKnownContract(normalizedTo, chainID);
  const targetName = contractInfo ? contractInfo.name : "";

  // If data is empty, return a simple transfer call
  if (cleanData.length === 0) {
    return {
      target: to,
      targetName: targetName,
      functionName: "unknown",
      rawData: "0x" + cleanData
    };
  }

  // Extract function selector (first 4 bytes)
  if (cleanData.length < 8) {
    return {
      target: to,
      targetName: targetName,
      functionName: "unknown",
      rawData: data
    };
  }

  const functionSelector = cleanData.slice(0, 8);

  // Try to identify the function from known selectors
  const functionInfo = knownFunctions[functionSelector];

  if (!functionInfo) {
    // If we can't identify the function, return the raw data
    return {
      target: to,
      targetName: targetName,
      functionName: "unknown",
      rawData: data
    };
  }

  // Parse the function arguments
  let parsedArgs;
  try {
    parsedArgs = parseArguments(functionInfo.abi, "0x" + cleanData);
  } catch (err) {
    // If we can't parse the arguments, return the raw data
    return {
      target: to,
      targetName: targetName,
      functionName: functionInfo.name,
      rawData: data
    };
  }

  // If this is a token function on a known contract where we have decimals, adjust the amount
  if (contractInfo && tokenFunctions[functionInfo.name] && contractInfo.decimals > 0 && parsedArgs["amount"]) {
    parsedArgs["amount"] = parseDecimals(parsedArgs["amount"], contractInfo.decimals);
  }

  // Check if any of the arguments are known contracts
  for (const [key, value] of Object.entries(parsedArgs)) {
    if (value && typeof value === 'string' && ethers.utils.isAddress(value)) {
      const knownContract = getKnownContract(value, chainID);
      if (knownContract) {
        parsedArgs[key] = `${value} (${knownContract.name} ✅)`;
      }
    }
  }

  // Check if this is a multicall contract and the function is a multicall function
  const isMulticallContract = multicallAddresses[chainID] && 
                             multicallAddresses[chainID][normalizedTo];
  
  const isMulticallFunction = functionInfo.name === "multiSend" || functionInfo.name === "aggregate3";

  if (isMulticallContract && isMulticallFunction) {
    // Parse subcalls - passing contract address, chain ID, and full function info
    try {
      const subcalls = parseMulticall(normalizedTo, chainID, functionInfo, parsedArgs, options);
      
      // Return with subcalls
      return {
        target: to,
        targetName: targetName,
        functionName: functionInfo.name,
        subCalls: subcalls
      };
    } catch (err) {
      console.error("Error parsing multicall:", err);
      // Return without subcalls in case of error
      return {
        target: to,
        targetName: targetName,
        functionName: functionInfo.name,
        parsedData: parsedArgs
      };
    }
  }

  // Regular function call
  return {
    target: to,
    targetName: targetName,
    functionName: functionInfo.name,
    parsedData: parsedArgs
  };
}

/**
 * Parses the amount and returns it as a string with the correct number of decimals
 * @param {BigNumber} amount - Amount as BigNumber
 * @param {number} decimals - Number of decimals
 * @returns {string} Formatted string
 */
export function parseDecimals(amount, decimals) {
  let amountStr = amount.toString();

  // Pad with leading zeros if needed
  if (amountStr.length <= decimals) {
    amountStr = '0'.repeat(decimals - amountStr.length + 1) + amountStr;
  }

  // Insert decimal point at the right position
  const decimalPos = amountStr.length - decimals;
  return amountStr.slice(0, decimalPos) + '.' + amountStr.slice(decimalPos);
}

/**
 * Decodes the function arguments from calldata
 * @param {Object} method - ABI method definition
 * @param {string} calldata - Transaction calldata
 * @returns {Object} Decoded arguments
 */
function parseArguments(method, calldata) {
  // Remove 0x prefix if present
  calldata = calldata.startsWith('0x') ? calldata : '0x' + calldata;

  // Create an interface with just this function
  const iface = new ethers.utils.Interface([method]);
  
  // Decode the call data
  const decodedData = iface.parseTransaction({ data: calldata });
  
  // Convert to a map
  const result = {};
  for (let i = 0; i < decodedData.args.length; i++) {
    let name = method.inputs[i]?.name || `arg${i}`;
    let arg = decodedData.args[i];

    // Handle byte arrays for better readability
    if (arg instanceof Uint8Array || (Array.isArray(arg) && arg.every(item => typeof item === 'number' && item < 256))) {
      result[name] = '0x' + Buffer.from(arg).toString('hex');
    } else {
      result[name] = arg;
    }
  }

  return result;
}

/**
 * Parses subcalls from a multicall function
 * @param {string} contractAddress - Multicall contract address
 * @param {number} chainID - Chain ID
 * @param {Object} functionInfo - Function information
 * @param {Object} args - Function arguments
 * @param {Object} options - Verification options
 * @returns {Array} Array of parsed subcalls
 */
function parseMulticall(contractAddress, chainID, functionInfo, args, options) {
  const subcalls = [];
  const normalizedAddress = contractAddress.toLowerCase();

  // Handle Safe Multisend contracts
  if (normalizedAddress === SAFE_MULTISEND_ADDRESS.toLowerCase() || 
      normalizedAddress === SAFE_MULTISEND_CALL_ONLY_141.toLowerCase()) {
    
    // For Safe Multisend contract, check the function signature
    if (functionInfo.signature === SAFE_MULTISEND_SIG) {
      // Parse multiSend calldata
      const hexData = args["transactions"];
      if (!hexData || !hexData.startsWith('0x')) {
        throw new Error("Invalid multiSend data format");
      }

      // Convert hex string to bytes
      const data = ethers.utils.arrayify(hexData);

      // multiSend data format: operation (1 byte) + to (20 bytes) + value (32 bytes) + dataLength (32 bytes) + data (variable)
      let pos = 0;
      while (pos < data.length) {
        // Ensure we have enough data for the fixed-size fields
        if (pos + 85 > data.length) break;

        // Extract operation (skip as it's always 0 per comment in Go code)
        pos++;

        // Extract to address
        const to = ethers.utils.getAddress('0x' + Buffer.from(data.slice(pos, pos + 20)).toString('hex'));
        pos += 20;

        // Skip extracting value for now
        pos += 32;

        // Extract data length
        const dataLengthBytes = data.slice(pos, pos + 32);
        pos += 32;
        const length = ethers.BigNumber.from(dataLengthBytes).toNumber();

        // Ensure we have enough data for the variable-size field
        if (pos + length > data.length) break;

        // Extract call data
        const callData = '0x' + Buffer.from(data.slice(pos, pos + length)).toString('hex');
        pos += length;

        // Parse the subcall
        try {
          const subcall = parseTransactionData(to, callData, chainID, options);
          subcalls.push(subcall);
        } catch (err) {
          console.error(`Error parsing subcall to ${to}:`, err);
        }
      }
    } else {
      throw new Error(`Unsupported function ${functionInfo.signature} for Safe Multisend contract`);
    }
  } else if (normalizedAddress === MULTICALL3_ADDRESS.toLowerCase()) {
    // For Multicall3 contract, check the function signature
    if (functionInfo.signature === AGGREGATE3_SIG) {
      // Parse aggregate3 calldata
      const calls = args["calls"];
      if (!Array.isArray(calls)) {
        throw new Error("Invalid aggregate3 data format");
      }

      for (const call of calls) {
        try {
          const subcall = parseTransactionData(call.target, call.callData, chainID, options);
          subcalls.push(subcall);
        } catch (err) {
          console.error(`Error parsing subcall to ${call.target}:`, err);
        }
      }
    } else {
      throw new Error(`Unsupported function ${functionInfo.signature} for Multicall3 contract`);
    }
  } else {
    // No generic parsing - return an error for unknown contracts
    throw new Error(`Unsupported multicall contract: ${contractAddress}`);
  }

  return subcalls;
}
